### PR TITLE
VB-4066 backend validations public booking mvp initial

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Move session validations out of SessionService into a new service.

## What is the intent behind these changes?

VB-4066 initial changes.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerSessionValidator.kt [hunk 1]
```diff
@@ -1,87 +1,27 @@
...
 import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.SessionTemplate
-import java.util.function.Predicate
+import uk.gov.justice.digital.hmpps.visitscheduler.utils.validators.SessionCategoryValidator
+import uk.gov.justice.digital.hmpps.visitscheduler.utils.validators.SessionIncentiveValidator
+import uk.gov.justice.digital.hmpps.visitscheduler.utils.validators.SessionLocationValidator
 
 const val LOCATION_NOT_PERMITTED = -10
 
 @Component
 class PrisonerSessionValidator(
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerSessionValidator.kt:13]: this is the original class and will eventually be removed once all methods are moved to the Service layer.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionLocationValidator.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    if (isSessionForAllPrisonerLocations(sessionTemplate)) {
+      return true
+    }
+
+    return if (sessionTemplate.includeLocationGroupType) {
+      sessionTemplate.permittedSessionLocationGroups.any { prisonerLevelMatcher.test(it, prisonerLevels) }
+    } else {
+      sessionTemplate.permittedSessionLocationGroups.none { prisonerLevelMatcher.test(it, prisonerLevels) }
+    }
+  }
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionLocationValidator.kt:39]: moved from SessionService

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionLocationValidatorTest.kt [hunk 3]
```diff
@@ -17,15 +17,12 @@ import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.SessionT
...
 import java.time.LocalDate
 
 @ExtendWith(MockitoExtension::class)
-class PrisonerSessionValidatorTest {
-  private val prisonerLevelMatcher = PrisonerLevelMatcher()
-  private val prisonerCategoryMatcher = PrisonerCategoryMatcher()
-  private val prisonerIncentiveLevelMatcher = PrisonerIncentiveLevelMatcher()
-  private val prisonerSessionValidator = PrisonerSessionValidator(prisonerLevelMatcher, prisonerCategoryMatcher, prisonerIncentiveLevelMatcher)
-
+class SessionLocationValidatorTest {
```
Comment (dhirajshettymoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionLocationValidatorTest.kt:24]: renamed this class as it was only doing Location tests.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerSessionValidationService.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  ): Boolean {
+    // if there are sessions with incentive levels, check if the session is available to prisoner else return true as this is available to all
+    val hasSessionsWithIncentiveGroups = sessionTemplates?.any { it.permittedSessionIncentiveLevelGroups.isNotEmpty() } ?: true
+    return if (hasSessionsWithIncentiveGroups) {
+      sessionIncentiveValidator.isValid(sessionTemplate, prisoner)
+    } else {
+      true
+    }
+  }
+}
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerSessionValidationService.kt:71]: looks nicer

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionCategoryValidator.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+@Component
+class SessionCategoryValidator(
+  private val prisonerCategoryMatcher: PrisonerCategoryMatcher,
+) {
+  companion object {
+    val LOG: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun isValid(
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionCategoryValidator.kt:18]: nice

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionIncentiveValidator.kt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+@Component
+class SessionIncentiveValidator(
+  private val incentiveLevelMatcher: PrisonerIncentiveLevelMatcher,
+) {
+  companion object {
+    val LOG: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun isValid(
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionIncentiveValidator.kt:19]: nice

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionLocationValidatorTest.kt (hunk 3 above)
Comment (AledWEvans) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionLocationValidatorTest.kt:24]: So why have you got a separation for your tests on location but not cat and incentive

Comment (dhirajshettymoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionLocationValidatorTest.kt:24]: this was an existing test which has been renamed and refactored. Will add category and incentive tests.

Comment (dhirajshettymoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionLocationValidatorTest.kt:24]: Added 2 new tests for category and incentives.

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**